### PR TITLE
ddl : fix column collate should use table's if it has when create table or alter table. (#13119 & #12327)

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve? 
cherry-pick #13119 and #12327 to release-2.1

### What is changed and how it works?
```
conflict file & code change
show.go
db_integration_test.go
db_test.go
mock.go
column_test.go
show_test.go
```

### Check List 

Tests 

 - Unit test
 - Integration test

Release note

 - fix column collate should use table's if it has when create table or alter table.

### CODE REVIEW

File: ddl/db_integration_test.go [hunk 1]
```diff
@@ -653,8 +653,8 @@ func (s *testIntegrationSuite) TestIgnoreColumnUTF8Charset(c *C) {
 	c.Assert(err, NotNil)
 	c.Assert(terror.ErrorEqual(err, table.ErrTruncatedWrongValueForField), IsTrue, Commentf("err %v", err))
 	tk.MustQuery("show create table t").Check(testkit.Rows("t CREATE TABLE `t` (\n" +
-		"  `a` varchar(10) CHARSET utf8 COLLATE utf8_bin DEFAULT NULL,\n" +
```
Comment (bb7133) [inline ddl/db_integration_test.go]: Why changing this keyword?

Comment (AilinKid) [reply ddl/db_integration_test.go]: Cause backport change the output in `show.go`

**Pull request merged**